### PR TITLE
register/handlers: remove context manager

### LINKED ISSUE #2242: memory leak in contextManager
Using pprof, we've tracked down a memory leak caused by an unbalanced set of calls to context() and release() on defaultContextManager. The global map from requests to contexts was growing indefinitely. See:

https://github.com/docker/distribution/blob/7dba427612198a11b161a27f9d40bb2dca1ccd20/registry/handlers/context.go

defaultContextManager is used in https://github.com/stevvooe/distribution/blob/7dba427612198a11b161a27f9d40bb2dca1ccd20/registry/handlers/app.go. In one case a release() is deferred, but in the other it is not.

One theory I had was that the request object used to be the same in both context() calls, but as of go 1.7 when the new Context management for request objects was introduced those request objects stopped being the same, so the deferred release was no longer sufficient to remove the requests from the global map. We have not confirmed this theory because in go 1.7 the context manager is no longer necessary anyway, so it's easier to just remove it.

### PR DESCRIPTION
**Pull request by stevvooe**
Signed-off-by: Stephen J Day <stephen.day@docker.com>

cc @vikstrous

Closes #2242

### CODE REVIEW

**Comment by vikstrous** [comment]
It's worth noting that the previous use of context had a memory leak because it was never deallocating request objects because they were all in a global map and a release() call was missing in one of the places where context() was called.

We'll test out this commit on distribution 2.5 and report back.

File: registry/handlers/app.go [hunk 1]
```diff
@@ -756,7 +754,7 @@ func (app *App) logError(context context.Context, errors errcode.Errors) {
 // context constructs the context object for the application. This only be
 // called once per request.
 func (app *App) context(w http.ResponseWriter, r *http.Request) *Context {
-	ctx := defaultContextManager.context(app, w, r)
+	ctx := r.Context()
```
Comment (whoshuu) [inline registry/handlers/app.go]: :+1:

**Comment by stevvooe** [comment]
Also, note that while this starts to address #1764, it does not remove the `handlers.Context` object. More must be done to address #1764.

File: registry/handlers/app.go [hunk 2]
```diff
@@ -649,6 +644,9 @@ func (app *App) dispatcher(dispatch dispatchFunc) http.Handler {
 		// Add username to request logging
 		context.Context = ctxu.WithLogger(context.Context, ctxu.GetLogger(context.Context, auth.UserNameKey))
 
+		// sync up context on the request.
+		r = r.WithContext(context)
```
Comment (stevvooe) [inline registry/handlers/app.go]: This line causes a crash in gorilla. Will need to remove this line if on an older version of gorilla.

**Comment by vikstrous** [comment]
It might be safer and easier for backporting to solve the memory leak this way: https://github.com/docker/distribution/pull/2240 and the context manager can be removed independently of this particular issue

**Comment by vikstrous** [comment]
Confirmed that this PR works and fixes the memory leak.

File: registry/handlers/app.go [hunk 3]
```diff
@@ -596,24 +596,19 @@ func (app *App) configureSecret(configuration *configuration.Configuration) {
 func (app *App) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	defer r.Body.Close() // ensure that request body is always closed.
 
-	// Instantiate an http context here so we can track the error codes
-	// returned by the request router.
-	ctx := defaultContextManager.context(app, w, r)
```
Comment (tt) [inline registry/handlers/app.go]: This is a breaking change. Before all contexts would derive from the app's context and now each request is a new context. Why is this not documented?

Comment (stevvooe) [inline registry/handlers/app.go]: App context provenance is maintained at https://github.com/docker/distribution/blob/master/registry/handlers/app.go#L637. What exactly did this break?

Comment (tt) [inline registry/handlers/app.go]: It broke a middleware.

The context used to initialize the middleware (via [`repository.InitFunc`](https://godoc.org/github.com/docker/distribution/registry/middleware/registry#InitFunc)) is related to the context I passed in via `handlers.NewApp(...)`.

However, the context received by the [`Repository`](https://godoc.org/github.com/docker/distribution#Repository) `Manifests` function is the request context and doesn't have access to elements in that original context.

It's true that there is a reference to the app (https://github.com/docker/distribution/blob/1935c8d50b6b9d3fa8a5fcdb860cd3241a35e8df/registry/handlers/app.go#L766) but how do you access it once the context has been wrapped inside another context?

Comment (stevvooe) [inline registry/handlers/app.go]: Were you accessing the `App` object directly? That is not supported.

**Pull request merged**